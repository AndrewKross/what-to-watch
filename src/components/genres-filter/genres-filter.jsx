import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { ActionCreator as DataActionCreator } from "../../reducer/data/data";
import { ActionCreator as AppStateActionCreator } from '../../reducer/app-state/app-state';
import { getCurrentGenre, getFilms } from '../../reducer/data/selectors';

const GenresFilter = ({ films, currentGenre, onGenreChangeClick }) => {
  const allGenres = films.map((film) => film.genre);
  const genres = [].concat(`All genres`, Array.from(new Set(allGenres)));
  return (
      <ul className="catalog__genres-list">
        {genres.map((genre) => (
            <li
                className={`catalog__genres-item ${currentGenre === genre
                  ? `catalog__genres-item--active` : ``}`}
                key={genre}
                onClick={() => { onGenreChangeClick(genre); }}
            >
              <a href="#" className="catalog__genres-link">
                {genre}
              </a>
            </li>
        ))}
      </ul>
  );
};

const mapStateToProps = (state) => ({
  currentGenre: getCurrentGenre(state),
  films: getFilms(state),
});

const mapDispatchToProps = (dispatch) => ({
  onGenreChangeClick: (genre) => {
    dispatch(DataActionCreator.changeCurrentGenre(genre));
    dispatch(AppStateActionCreator.resetShowingFilmsCount());
  },
});

GenresFilter.propTypes = {
  currentGenre: PropTypes.string.isRequired,
  onGenreChangeClick: PropTypes.func.isRequired,
  films: PropTypes.array.isRequired,
};

export { GenresFilter as GenresFilterComponent };
export default connect(mapStateToProps, mapDispatchToProps)(GenresFilter);
