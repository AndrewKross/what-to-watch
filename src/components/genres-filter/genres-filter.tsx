import * as React from 'react';
import { connect } from 'react-redux';
import { ActionCreator as AppStateActionCreator } from '../../reducer/app-state/app-state';
import { getFilms } from '../../reducer/data/selectors';
import { getCurrentGenre } from '../../reducer/app-state/selectors';
import { Film } from '../../types';

interface Props {
  films: Film[];
  currentGenre: string;
  onGenreChangeClick: (genre: string) => void;
}

const GenresFilter: React.FunctionComponent<Props> = ({
  films, currentGenre, onGenreChangeClick
}: Props) => {
  const allGenres = films.map((film) => film.genre);
  const genres = [].concat(`All genres`, Array.from(new Set(allGenres)));
  return (
    <ul className="catalog__genres-list">
      {genres.map((genre) => (
        <li
          className={`catalog__genres-item ${currentGenre === genre
            ? `catalog__genres-item--active` : ``}`}
          key={genre}
          onClick={() => { if (currentGenre !== genre) { onGenreChangeClick(genre); } }}
        >
          <a href="#" className="catalog__genres-link">
            {genre}
          </a>
        </li>
      ))}
    </ul>
  );
};

const mapStateToProps = (state) => ( {
  currentGenre: getCurrentGenre(state),
  films: getFilms(state),
} );

const mapDispatchToProps = (dispatch) => ( {
  onGenreChangeClick: (genre) => {
    dispatch(AppStateActionCreator.changeCurrentGenre(genre));
    dispatch(AppStateActionCreator.resetShowingFilmsCount());
  },
} );

export { GenresFilter as GenresFilterComponent };
export default connect(mapStateToProps, mapDispatchToProps)(GenresFilter);
