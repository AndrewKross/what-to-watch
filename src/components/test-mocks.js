export const PromoFilmData = {
  title: `The Grand Budapest Hotel`,
  genre: `Drama`,
  date: `2014`
};

const GENRE = `drama`;
const RELEASE = `2020`;
const RATING = 9;
const VOTES = 228;
const DIRECTOR = `Otto Bathurst`;
const PREVIEW = `https://cdn.videvo.net/videvo_files/video/free/2019-09/small_watermarked/190828_07_MarinaBayatNightDrone_UHD_02_preview.webm`;
const RUN_TIME = 140;

const DESCRRIPTION = `In the 1930s, the Grand Budapest Hotel is a popular European ski resort,
presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's
friend and protege. Gustave prides himself on providing first-class service to the hotel's guests,
including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's
lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief
suspect in her murder.`;

const TITLES = [
  `Fantastic Beasts: The Crimes of Grindelwald`,
  `Bohemian Rhapsody`,
  `Macbeth`,
  `Aviator`,
  `We need to talk about Kevin`,
  `What We Do in the Shadows`,
  `Revenant`,
  `Johnny English`
];

const COVERS = [
  `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  `img/bohemian-rhapsody.jpg`,
  `img/macbeth.jpg`,
  `img/aviator.jpg`,
  `img/we-need-to-talk-about-kevin.jpg`,
  `img/what-we-do-in-the-shadows.jpg`,
  `img/revenant.jpg`,
  `img/johnny-english.jpg`
];

const ACTORS = [
  `Cillian Murphy`,
  `Sam Neill`,
  `Helen McCrory`,
  `Paul Anderson`,
  `Annabelle Wallis`,
  `Iddo Goldberg`
];

const review = {
  id: `0`,
  text: `I didn't find it amusing, and while I can appreciate the creativity,
  it's an hour and 40 minutes I wish I could take back.`,
  rating: 8,
  userName: `Kate Muir`,
  date: new Date()
};

export const films = TITLES.map((title, index) => {
  return ({
    id: index.toString(),
    title,
    cover: COVERS[index],
    poster: COVERS[index],
    preview: PREVIEW,
    genre: GENRE,
    release: RELEASE,
    rating: RATING,
    votes: VOTES,
    description: DESCRRIPTION,
    director: DIRECTOR,
    actors: ACTORS,
    runTime: RUN_TIME,
    reviews: [review]
  });
});

export const TABS_DATA = [
  `overview`,
  `details`,
  `reviews`
];
