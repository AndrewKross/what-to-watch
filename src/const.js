export const NUMBER_OF_SIMILAR_FILMS = 4;
export const PREVIEW_DELAY = 1000;
export const REVIEWS_IN_COLUMN_COUNT = 3;
export const FILMS_ON_START_SCREEN = 8;
export const FILMS_ON_SHOW_MORE_BUTTON = 20;
export const REVIEW_DATE_HUMAN_FORMAT = `MMMM D, YYYY`;
export const REVIEW_DATE_SERVICE_FORMAT = `YYYY-MM-DD`;

export const RatingRange = {
  MIN: 0,
  MAX_BAD: 3,
  MAX_NORMAL: 5,
  MAX_GOOD: 8,
  MAX: 10,
};

export const RatingGrade = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`,
};

export const AppRoute = {
  MAIN: `/`,
  FILM: `/film/`,
  PLAYER: `/player/`,
};

export const GENRES = [
  `All genres`,
  `Comedies`,
  `Crime`,
  `Documentary`,
  `Dramas`,
  `Horror`,
  `Kids & Family`,
  `Romance`,
  `Sci-Fi`,
  `Thrillers`,
];

export const GenresForFilms = {
  [`Comedies`]: `comedy`,
  [`Crime`]: `crime`,
  [`Documentary`]: `documentary`,
  [`Dramas`]: `drama`,
  [`Horror`]: `horror`,
  [`Kids & Family`]: `kids & family`,
  [`Romance`]: `romance`,
  [`Sci-Fi`]: `sci-fi`,
  [`Thrillers`]: `thriller`,
};

export const TabsNames = {
  OVERVIEW: `overview`,
  DETAILS: `details`,
  REVIEWS: `reviews`,
};

export const TabsData = [{
  name: TabsNames.OVERVIEW,
  content: `Overview`,
},
{
  name: TabsNames.DETAILS,
  content: `Details`,
},
{
  name: TabsNames.REVIEWS,
  content: `Reviews`,
},
];
