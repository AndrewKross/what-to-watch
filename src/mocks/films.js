import { nanoid } from "nanoid";
import {
  getRandomIntegerNumber,
  getRandomArrayItem,
  getRandomArrayItems,
} from "../utils/common.js";
import { reviews } from "./reviews.js";

const ID_LENGTH = 8;

const TITLES = [
  `Fantastic Beasts: The Crimes of Grindelwald`,
  `Bohemian Rhapsody`,
  `Macbeth`,
  `Aviator`,
  `We need to talk about Kevin`,
  `What We Do in the Shadows`,
  `Revenant`,
  `Johnny English`,
];

const COVERS = [
  `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  `img/bohemian-rhapsody.jpg`,
  `img/macbeth.jpg`,
  `img/aviator.jpg`,
  `img/we-need-to-talk-about-kevin.jpg`,
  `img/what-we-do-in-the-shadows.jpg`,
  `img/revenant.jpg`,
  `img/johnny-english.jpg`,
];

const PREVIEWS = [
  `https://cdn.videvo.net/videvo_files/video/free/2019-09/small_watermarked/190828_07_MarinaBayatNightDrone_UHD_02_preview.webm`, `https://www.videvo.net/videvo_files/converted/2018_07/preview/180607_A_064.mp429860.webm`, `https://www.videvo.net/videvo_files/converted/2018_01/preview/171124_H1_006.mp491073.webm`, `https://www.videvo.net/videvo_files/converted/2016_11/preview/GOPR6239_1.mov34724.webm`, `https://cdn.videvo.net/videvo_files/video/free/2019-12/small_watermarked/190915_B_01_Timelapse%20Danang_01_preview.webm`,
];

const GENRES = [
  `comedy`,
  `drama`,
  `melodrama`,
  `horror`,
  `crime`,
  `documentary`,
  `romance`,
  `thriller`,
];

const DESCRRIPTION = `In the 1930s, the Grand Budapest Hotel is a popular European ski resort,
presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's
friend and protege. Gustave prides himself on providing first-class service to the hotel's guests,
including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's
lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief
suspect in her murder.`;

const DIRECTORS = [
  `Otto Bathurst`,
  `Tom Harper`,
  `Colm McCarthy`,
  `Tim Mielants`,
  `David Caffrey`,
];

const ACTORS = [
  `Cillian Murphy`,
  `Sam Neill`,
  `Helen McCrory`,
  `Paul Anderson`,
  `Annabelle Wallis`,
  `Iddo Goldberg`,
  `Sophie Rundle`,
  `Joe Cole`,
  `Finn Cole`,
];

const ReleaseRange = {
  MIN: 1980,
  MAX: 2020,
};

const RatingRange = {
  MIN: 0,
  MAX: 10,
};

const ratingsCountRange = {
  MIN: 0,
  MAX: 1000,
};

const RunTimeRange = {
  MIN: 20,
  MAX: 240,
};

export const filmsData = TITLES.map((title, index) => {
  return {
    id: nanoid(ID_LENGTH),
    title,
    cover: COVERS[index],
    poster: COVERS[index],
    preview: getRandomArrayItem(PREVIEWS),
    genre: getRandomArrayItem(GENRES),
    release: getRandomIntegerNumber(
      ReleaseRange.MIN,
      ReleaseRange.MAX
    ).toString(),
    rating: +(Math.random() * RatingRange.MAX).toPrecision(2),
    ratingsCount: getRandomIntegerNumber(
      ratingsCountRange.MIN,
      ratingsCountRange.MAX
    ),
    description: DESCRRIPTION,
    director: getRandomArrayItem(DIRECTORS),
    actors: getRandomArrayItems(ACTORS),
    runTime: getRandomIntegerNumber(RunTimeRange.MIN, RunTimeRange.MAX),
    reviews,
  };
});
