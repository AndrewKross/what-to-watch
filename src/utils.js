export const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
};

export const getRatingLevel = (rating) => {
  rating = parseInt(rating, 10);
  if (rating >= 0 && rating < 3) {
    return `Bad`;
  }
  if (rating < 5) {
    return `Normal`;
  }
  if (rating < 8) {
    return `Good`;
  }
  if (rating < 10) {
    return `Very good`;
  }
  if (rating === 10) {
    return `Awesome`;
  }
  return null;
};
