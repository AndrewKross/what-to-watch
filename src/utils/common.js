import moment from "moment";

export const getRatingLevel = (rating) => {
  rating = parseInt(rating, 10);
  if (rating >= 0 && rating < 3) {
    return `Bad`;
  }
  if (rating < 5) {
    return `Normal`;
  }
  if (rating < 8) {
    return `Good`;
  }
  if (rating < 10) {
    return `Very good`;
  }
  if (rating === 10) {
    return `Awesome`;
  }
  return null;
};

export const getRandomIntegerNumber = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);

  return Math.floor(Math.random() * (max - min + 1)) + min;
};

export const getRandomArrayItem = (elements) => {
  const randomIndex = getRandomIntegerNumber(0, elements.length - 1);

  return elements[randomIndex];
};

const shuffleArray = (array) => {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));

    [array[i], array[j]] = [array[j], array[i]];
  }

  return array;
};

export const getRandomArrayItems = (array) => {
  const shuffledArray = shuffleArray(array.slice());

  return shuffledArray.slice(0, getRandomIntegerNumber(0, array.length));
};

export const getFormatedRunTime = (duration) => {
  const durationInMinutes = moment.duration(duration, `minutes`);

  return `${durationInMinutes.hours()}h ${durationInMinutes.minutes()}m`;
};
