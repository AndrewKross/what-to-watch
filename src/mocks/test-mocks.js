import { NameSpace } from '../reducer/reducer';
import { DEFAULT_GENRE, LoadingStatus } from '../const';

const RELEASE = 2020;
const RATING = 9;
const RATINGS_COUNT = 228;
const DIRECTOR = `Otto Bathurst`;
const PREVIEW = `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`;
const RUN_TIME = 140;
const COMMENTS_COUNT = 6;

const DESCRIPTION = `In the 1930s, the Grand Budapest Hotel is a popular European ski resort,
presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's
friend and protege. Gustave prides himself on providing first-class service to the hotel's guests,
including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's
lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief
suspect in her murder.`;

const TITLES = [
  `Fantastic Beasts: The Crimes of Grindelwald`,
  `Bohemian Rhapsody`,
  `Macbeth`,
  `Aviator`,
  `We need to talk about Kevin`,
  `What We Do in the Shadows`,
  `Revenant`,
  `Johnny English`,
];

const COVERS = [
  `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  `img/bohemian-rhapsody.jpg`,
  `img/macbeth.jpg`,
  `img/aviator.jpg`,
  `img/we-need-to-talk-about-kevin.jpg`,
  `img/what-we-do-in-the-shadows.jpg`,
  `img/revenant.jpg`,
  `img/johnny-english.jpg`,
];

const ACTORS = [
  `Cillian Murphy`,
  `Sam Neill`,
  `Helen McCrory`,
  `Paul Anderson`,
  `Annabelle Wallis`,
  `Iddo Goldberg`,
];

export const GENRES = [
  `all genres`,
  `comedy`,
  `drama`,
  `melodrama`,
  `horror`,
  `crime`,
  `documentary`,
  `romance`,
  `thriller`,
];

export const TABS_DATA = [
  `overview`,
  `details`,
  `reviews`,
];

const comment = {
  id: 0,
  text: `I didn't find it amusing, and while I can appreciate the creativity,
  it's an hour and 40 minutes I wish I could take back.`,
  rating: 8,
  userName: `Kate Muir`,
  date: new Date(`December 25, 2019 01:00:00`),
};

export const comments = Array(COMMENTS_COUNT)
  .fill(``)
  .map(() => comment);

export const films = TITLES.map((title, index) => {
  return ({
    id: index,
    title,
    cover: COVERS[index],
    posterImage: COVERS[index],
    previewImage: COVERS[index],
    backgroundImage: COVERS[index],
    previewVideo: PREVIEW,
    videoMain: PREVIEW,
    genre: GENRES[index + 1] || GENRES[1],
    released: RELEASE,
    rating: RATING,
    ratingsCount: RATINGS_COUNT,
    description: DESCRIPTION,
    director: DIRECTOR,
    starring: ACTORS,
    runTime: RUN_TIME,
    isFavorite: false,
    backgroundColor: `black`,
  });
});

export const userData = {
  id: 1,
  email: `Andrew@gmail.com`,
  name: `Andrew`,
  avatar: `img/1.png`,
};

export const loginData = {
  email: `Andrew@gmail.com`,
  password: `1234`,
};

export const reviewToPost = {
  rating: 8,
  comment: `Great!`,
};

export const defaultMockStore = {
  [NameSpace.DATA]: {
    films,
    comments,
    isFilmsLoaded: true,
    isCommentsLoaded: true,
  },
  [NameSpace.APP_STATE]: {
    filmsOnScreen: 8,
    currentGenre: DEFAULT_GENRE,
  },
  [NameSpace.USER]: {
    isAuthorized: true,
    userInfo: userData,
    reviewLoadingStatus: LoadingStatus.PENDING,
    authorizationLoadingStatus: LoadingStatus.PENDING,
  },
};
