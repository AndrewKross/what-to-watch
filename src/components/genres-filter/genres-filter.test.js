import React from "react";
import renderer from "react-test-renderer";
import { GenresFilterComponent } from "./genres-filter.jsx";
import { DEFAULT_GENRE } from '../../const';
import { films } from '../../mocks/test-mocks';
import { noop } from '../../utils/common';

test(`GenresFilter should render correctly`, () => {
  const tree = renderer
    .create(
      <GenresFilterComponent
          currentGenre={DEFAULT_GENRE}
          films={films}
          onGenreChangeClick={noop}
      />, {
        createNodeMock: () => ({}),
      },
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
