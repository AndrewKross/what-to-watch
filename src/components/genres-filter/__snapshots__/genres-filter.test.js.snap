// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenresFilter should render correctly 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item catalog__genres-item--active"
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      All genres
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      comedy
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      drama
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      melodrama
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      horror
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      crime
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      documentary
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      romance
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      thriller
    </a>
  </li>
</ul>
`;
